Correct grammar in README.md (#17)
KhartesViewer/khartes: README.md

### 2-minute video

The video below gives a quick, 2-minute overview of khartes.

(If you click on the image below, you will be taken to vimeo.com to watch the video)

[![Watch the video](https://i.vimeocdn.com/video/1703962064-7e3db7142e72fe9b85887c02bd8902a65a1a6ef9c644e3e12bb5485b3519ece3-d_640)](https://vimeo.com/849799186)

### 30-minute video

This next video provides a more extensive introduction to khartes.
Note that it begins with a 60-second
intro sequence which contains no narration, but which quickly highlights some of khartes' features.
After the intro, the video follows a more traditional format, with a voice-over and a demo.
The entire video is about 30 minutes long.  There is no closed captioning, but the script for the video
can be found in the file demo1_script.txt.

(If you click on the image below, you will be taken to vimeo.com to watch the video)

[![Watch the video](https://i.vimeocdn.com/video/1670955201-81a75343b71db9c84b6b4275e3447c943d2128ab8b921a822051046e83db0c96-d_640)](https://vimeo.com/827515595)

## Installation

You should be able to run khartes simply by
cloning the repository, making sure you have the proper dependencies

> (see [anaconda_installation.txt](anaconda_installation.txt) for a list), and then typing `python khartes.py`.

Reviewer: Thanks for adding the link!

same intersection line is drawn in red, so that when you are looking at the green window,
you know where the red slice crosses it.

### Navigation by mouse
#### Navigation by mouse

To navigate within the data volume, simply hold down the left mouse button while inside one
of the Data Slice windows, and drag the slice.  The two other slices will also change, to ensure that
the mutual intersection point of the three slices remains in the middle of the crosshairs
of all 3 slices.
of all three slices.

You can use the mouse scroll wheel while in any of the Data Slice windows to zoom in or out.
The crosshairs are always the center of the zoom; the mutual intersection point of the three
slice planes does not change during zooming.

### Navigation by keyboard
#### Navigation by keyboard

> You can use the four arrow key to move the current slice (the one where the mouse cursor is

Reviewer: Nice catch!

> You can use the four arrow keys to move the current slice (the one where the mouse cursor is

Reviewer: Nice catch!

> You can use the four arrow keys to move the current slice (the one where the mouse cursor is

Reviewer: Nice catch!

> that their mutual intersection point remains in the centers of all three pairs of crosshairs.

Reviewer: Another nice catch!

If your keyboard offers auto-repeat, you can hold down an arrow key, or a page-up/down key, and
watch as the current slice slowly moves in the indicated direction.

> The `a`, `s`, `w`, and `d` keys behave the same as the arrow keys, and `e` and `c` behave the

Reviewer: And another!  This one is especially significant; before your fix, it gave incorrect instructions to the user.

At this point you can begin adding points (called nodes) to define
the fragment.  To add a node, position the cursor in one of the
Data Slices, hold down the shift key (so the cursor turns into

> crosshairs), and press the left mouse button.

Reviewer: Nice improvement!

Also, the video does not describe the user's key presses and button
presses.

Here are some general user-interface aspects that you should
have gathered from the video:

* Each node that the user creates is added to a surface (the fragment);
this surface is displayed in the map view under the Fragment View.  Of course, if
the new nodes form a line, rather than a surface,

> it is shown in the Fragment View.  Once the user creates nodes

Reviewer: ```suggestion
* Each node that the user creates is added to a surface (the fragment);
a map view of this surface is displayed in the Fragment View window.  Of course, if
all the nodes lie in a straight line,
then a line, rather than a surface, is shown in the Fragment View.  Once the user creates nodes
```

Author: ```suggestion
* Each node that the user creates is added to a surface (the fragment);
a map view of this surface is displayed in the Fragment View window.
If all the nodes lie in a straight line,
then the Fragment View will display a line rather than a surface.  Once the user creates nodes
```
Sure. However, let's keep it short and to the point; too many commas make it hard to comprehend.
I have re-arranged it and dropped "Of course" as it adds unnecessary detail.

Author: Unresolving comment. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

tutorial, but you need to be alerted to it now, in case you
run into it by accident.

### Normal Mode and Add-Node Mode

Khartes has two modes of user interaction in the data views:
"normal" mode, and "add-node" mode.  In normal mode, the
cursor usually looks like a hand (unless it is near a node), and
to add a node, you need to hold down the shift key, so the cursor
turns into crosshairs, before pressing the left mouse button.

In "add-node" mode, things are reversed.  The cursor always
takes the form of crosshairs, and to add a node, you simply need to
press the left mouse button.

There are two ways to toggle between the two modes.  One way is to
click on the button, to the right of the `Files` menu, that contains
a crosshairs icon.  When this button is enabled, khartes is in

> normal mode, and when it is disabled, khartes is in add-node mode.

Reviewer: I'm not sure what to do here.  My original "light" and "dark" are incorrect, if the user is in dark mode.  But I don't know whether users understand "enabled" and "disabled" in this context.

Author: ```diff
- a picture of crosshairs.  When this button is light, khartes is in
- normal mode, and when it is dark, khartes is in add-node mode.
+ a crosshairs icon.  When this button is not selected, khartes is in
+ normal mode, and when it is selected, khartes is in add-node mode.
```
I have modified it to use "selected" instead. (fyi suggesting changes is not handled, will commit it manually)
Also, let me know if I need to switch them around. I assume `on = normal mode`.

(edited to include below modification)

Reviewer: Looks good, except selected = add-node mode.

Author: Thanks. I did think that'd be the case.

mode you are in by seeing whether the cursor afterwards is a hand (normal mode)
or crosshairs (add-node mode).

This topic needed to be introduced now, because you might, by accident,
hit the shift button, unintentionally toggling between modes.
Now that you understand the two modes, you won't be mystified if this happens.

### Choosing a region for segmentation

Before you start creating a fragment, you should make sure that there
is a region in your high-resolution data volume that is possible to segment.

What you are looking for is a region where the papyrus sheets, as
seen in the top Data Slice, are not too tilted (you want an angle
flatter than 45 degrees or so).  This region doesn't have to be very large,
but if you can't find even a small area, then you may need to create
another high-resolution data volume.

### Building and modifying the segment

At this point you are ready to build the segment.
As you saw in the video, you do this by simply adding nodes to the
Data Slices.  Notice that **the order you add nodes does not matter**;
if you feel your nodes are too far apart, simply add more.

But what do you do if you discover that a node is in the wrong

> place? You have a couple of options.

Reviewer: ```suggestion
place?  You have a couple of options.
```

Author: Will do.

Author: Unresolving comment. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

#### Deleting a node

It is very easy to delete a node.  Simply move the cursor over
the node, so that the node turns cyan.
At this point, simply hit the "Delete" or "Backspace" key (either will work).

> But bear in mind **deleting a node will remove it permanently**.

Reviewer: ```suggestion
But bear in mind that **deleting a node will remove it permanently**;
there is no "undo" in khartes!
```

Author: ```suggestion
But bear in mind that **deleting a node will remove it permanently**;
it can not be recovered.
```
I have modified it to avoid the 'double negative' "no" and "un-".
Also, let's remove the exclamation mark, as it is exhaustive; the fact it is permanent is already emphasized using **bold** formatting.

Reviewer: I agree with the added "that", but I do want to reiterate wherever possible that khartes really does not have an "undo" mode; this is a handy place to do that.  I agree that the exclamation mark is excessive.

Author: ```suggestion
But bear in mind that **deleting a node will remove it permanently**;
it can not be recovered, as there is currently no undo action implemented.
```

I have added a mention of the undo action, but I have kept the mention of recovery as it helps to keep it in line with the rest of the tutorial.

I should clarify that I also removed the mention of the undo mode because it is already mentioned in https://github.com/KhartesViewer/khartes#things-to-fix , so it may only be temporary. Removing it avoids having to maintain such details. It could alternatively be mentioned once, in bold, at the start of the tutorial.

The added "that" came from your suggested change, but it does look fine.

Reviewer: That looks good

page-up or page-down in the Fragment View).

**The key capability of khartes is that as you move nodes, the data image in the
Fragment View updates interactively, so you can instantly judge the quality of
your segmentation.**

#### What the cursor is telling you

As you may have noticed, the cursor often changes its appearance,
depending on where it is in the slice.  The rule is that the shape

> of the cursor tells you what will happen when you press the left mouse button.

Reviewer: Nice improvement!

a new node will be created at the center of the crosshairs.

* **Diagonal two-way arrow**: You will only see this while using the TIFF Loader to create a new

> data volume.  This shape indicates that the cursor is near a corner of the

Reviewer: Good catch!

* Have a fragment that is active (one that is accepting new nodes)

* Make sure you are in add-node mode (the cursor is shaped like a crosshair)

* You need to be working in one of the two top windows in the Data Slices area.

At this point, you only need to choose where to begin, and in which
direction to go.  If the cursor is near
an existing node (the node has turned cyan), the auto-segmentation will start

> at that node.  Otherwise, it will start at the location of the cursor.

Reviewer: Good!

To automatically create new nodes to the right of the first one, press

> the right bracket or brace key (`]` or `}`); either key will work.  To create new nodes to the

Reviewer: ```suggestion
To automatically create new nodes at and to the right of the starting location,
or to the right of the selected node, press
the right bracket or brace key (`]` or `}`); either key will work.  To create new nodes to the
```

Reviewer: I'm open to other suggestions here; the main thing to keep in mind is that the cursor may be near an existing node, in which case the new nodes are created to the right of the existing node; or the cursor may not be near an existing node, in which case a new node is created at the cursor location, and at locations to the right.

Author: ```diff
+ The auto-segmentation algorithm will automatically create new nodes, going as far as [...]

- At this point, you only need to choose where to begin, and in which direction to go. [...]
- Otherwise, it will start at the location of the cursor.
+ You only need to choose where to begin, and in which direction to go. [...]
+ Otherwise, it will create a node at the location of the cursor and continue from there.

- To automatically create new nodes to the right of the first one, press the right bracket or brace key (] or }); either key will work.
- To create new nodes to the left, press the left bracket or brace key ([] or {).
+ To start and head to the right, press the right bracket or brace key (] or }).
+ To start and head to the left, press the left bracket or brace key ([] or {) instead.

- The auto-segmentation algorithm will go as far as [...]
```
I have modified it to explain the process first and then give instructions.
(If this change is okay, will commit manually.)

Author: Unresolving comment. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

It is a good idea, until you get a feeling for its capabilities,
to work on a fairly zoomed-in
area, in order to limit how far the algorithm goes.  This means you
will not have to delete as many nodes if the algorithm goes
astray (as it often does).

> Note that the algorithm is quite fast; it often adds new nodes within

Reviewer: ```suggestion
Note that the algorithm is quite fast; it often adds the new nodes within
```

Author: Will do.

Author: Unresolving comment. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

a couple of seconds (longer for a larger surface, due to khartes updating
the surface to account for the new nodes).  If nothing happens within 10 to 15 seconds,

> check to make sure that you meet the prerequisites above.

Reviewer: ```suggestion
check to make sure that you have met the prerequisites above.
```

Author: Sure, my mistake. I would like to avoid mixed past and present tense, but upon further thought this is fine.

Author: Unresolving comment. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

* **shift**: As mentioned in the "Normal Mode and Add-Node Mode" section above, quickly pressing and releasing the
shift key will toggle between "normal mode" and "add-node mode".  See the section below
entitled "Advanced Topic: Control Area: Settings" for information on how to change this

> behavior.

Reviewer: Another good catch!  But you have discovered so many typos that henceforth I will not mention each one.  So a collective "good catch" for all the typos you have discovered but that I haven't called out!

* **delete** and **backspace**: If you place the cursor over a node, so that the node turns cyan,
you can delete the node by pressing either of these two keys.  As mentioned above, deleting a node

> will delete it permanently, so think carefully before doing this.

Reviewer: ```suggestion
will delete it permanently (khartes does not have an "undo" function),
so think carefully before doing this.
```

Author: ```diff
- you can delete the node by pressing either of these two keys.  As mentioned above, khartes does
- not have an "undo" function, so think carefully before doing this.
+ you can delete the node by pressing either of these two keys.  As mentioned above, deleted nodes
+ can not be recovered, so think carefully before doing this.
```

See https://github.com/KhartesViewer/khartes/pull/17#discussion_r1381921243 .

Reviewer: Same comment as before; I want to emphasize the general lack of an undo function.

Author: I would suggest to avoid duplicate content. This refers back to the first mention, in which it can already be handled.
See https://github.com/KhartesViewer/khartes/pull/17#discussion_r1382369773

Reviewer: That seems fine.

on its side (in the Control Area, find the Volume tab, and
change the direction of your volume from "Y" to "X").

**Step 1**: Start in an easy area of your sheet, picking some points
on the inline (top window) and crossline (middle window) Data Slices.
This will create a diamond-shaped area in the Fragment Viewer.
Make sure you are happy with what you see
before expanding.

**Step 2**: Expand by alternating directions.  Use the Fragment Viewer
to move to a different area on the fragment and create nodes on the inline
slice.  Then create nodes on the crossline slice.  You can also add
nodes to the bottom slice; these act like contour lines on a map.

**Hint for step 2**: Before you start adding new nodes onto the
inline or crossline slice,
look in the Fragment Viewer to see if there are any existing nodes near
the line you are working on.
If there are, and it is feasible, move these existing nodes
onto the line.  This is to avoid the situation where a node on the line
and a node next to the line end up too close to each other, which can
cause undesirable waviness in the fragment.

<img src="images/good start.JPG" width="800" />

> *Example of a good start.* A couple of inline slices (the horizontal lines

Reviewer: ```suggestion
*Example of a good start.*   A couple of inline slices (the horizontal lines
```

Author: ```suggestion
*Example of a good start.*  A couple of inline slices (the horizontal lines
```
It only needs one more space, not two. There was already one there.

Author: Unresolving comment. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

in the Fragment View) and a crossline slice (the vertical line) have been
interpreted.  Nodes near the lines have been moved onto the lines, to
maintain good node spacing.
Some contour points have been added to the bottom slice as well.
The horizontal fibers are continuous,
which is a sign that the segmentation has been done correctly (see Step 3).
The dark spot in the upper right quadrant is
due to a lack of nodes to constrain the interpolation.

**Step 3**: Pause, verify, repair.  The most important criterion for
a good fragment is that the horizontal fibers (as seen in the Fragment View)
are continuous, since the horizontal fibers (also called the circumferential fibers)
are the ones that are most likely to contain text.
Where horizontal and vertical fibers cross, try
to make sure that the horizontal fibers are the ones that are the most visible

![sheet_skip](images/sheet_skip.JPG)

> ***This is bad!*** The horizontal fibers are not continuous.  This needs to be repaired by

Reviewer: ```suggestion
***This is bad!***  The horizontal fibers are not continuous.  This needs to be repaired by
```

Author: Will do.

Author: Unresolving comment. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

*The surface has been repaired;* horizontal fibers are now continuous.  The inline and crossline slices show

> the location of the original (magenta) and repaired (cyan) surfaces.  Note that these overlap in one half

Author: I assume these colours are the same regardless of the user being in light or dark mode.
Let me know if this is correct.

Reviewer: Yes.  The colors (magenta and cyan) were set by the user, and so do not depend on light/dark mode.  But I'm glad you checked.

of the slice, but diverge in the other half.

### Workflow notes

> **Save often.** You simply need to type Ctrl-S to save your work; try to remember to do this whenever

Reviewer: ```suggestion
**Save often.**  You simply need to type Ctrl-S to save your work; try to remember to do this whenever
```

Author: Will do.

Author: Unresolving comment. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

takes.  If a node is surrounded by large triangles that cover most
of the data volume, each change may require several seconds to recompute,
meaning that khartes will no longer feel interactive.  You can prevent this problem
by keeping your triangles regular and local.

So, when segmenting, start in the center of a fragment
and work your way out, keeping a fairly regular mesh, instead

> of trying to create a huge surface.  This practice

Reviewer: ```suggestion
of immediately trying to create a huge surface.  This practice
```

Author: Will do.

Author: Unresolving comment. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

You have already visited this tab, in order to create new fragments;
now you will have a chance to study it in detail.

> The first thing to notice is than one of the rows is highlighted with a

Reviewer: ```suggestion
The first thing to notice is that one of the rows is highlighted with a
```

Author: Will do. No idea what happened here.

Author: Unresolving comment. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

it does allow you to add dots to the Fragment View to highlight features of interest.

These dots are actually the nodes of a fragment whose only purpose

> is to provide annotation.  Here are the steps to add an annotation to

Reviewer: ```suggestion
is to provide annotation.  Here are the steps to add annotation to
```

Author: Will do. I see it is more consistent with this section.

Author: Unresolving comment. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

8. If you want annotation in more than one color, just create more annotation
fragments, one per color.  Make sure they are all `Active`.

### Exporting fragments

> Khartes allows you to export your fragments to [MeshLab](https://www.meshlab.net/),

Reviewer: I'm curious to know why you made this change?  I'm not necessarily against it, just curious to know the reason.  Same with the change to the volume-cartographer link below, and to the subsequent mentions of MeshLab.

Author: I assume these are titles/names of products, not their components.
Products are common 'terms', whereas components are usually used in a computing-related context, i.e. 'code'.
Unless the common name is explicitly `formatted like this`, it is better not to do so.

Author: Marking comment as resolved. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

created and maintained by the [Visual Computing Lab](https://vcg.isti.cnr.it/) of [CNR-ISTI](https://www.isti.cnr.it/en/) (Italy),
as well as to
`vc_render` and `vc_layers_from_ppm`, which are two components of [volume-cartographer](https://github.com/educelab/volume-cartographer),
created and maintained by the [Digital Restoration Initiative](https://www2.cs.uky.edu/dri/), part of
[EduceLab](https://educelab.engr.uky.edu/) at the [University of Kentucky](https://uky.edu/) (USA).

To export your fragment:

1. Make sure your fragment is active, that is, that it is visible
in the right-hand window.
2. In the File menu, select `Export file as mesh...`.
3. After you press the `Save` button, you will be asked to specify

> an "infill spacing".  The default value should be fine.

Reviewer: I like this wording!

that you want to give the `.ppm` file.  The `.ppm` file is needed by
`vc_layers_from_ppm`.

So, the command you want to type will look something like:
```
vc_render -v [your volpkg directory] --input-mesh [your .obj file] --output-ppm [the name of the ppm file you want to create]
```

> You might need to use `--volume` to specify your volume as well, if your `.volpkg` has more than one.

Reviewer: ```suggestion
You might need to use `--volume` to specify your volume as well, if your volpkg has more than one.
```

Reviewer: I'm using volpkg as a noun here, rather than as a file extension.  But I will think about this some more; I might end up changing back to your suggestion.

Author: Sure.

Author: Could you let me know what your choice is here, please?
If you want to change it back to a **noun**, you can also suggest to make the change elsewhere.
(depends on your interpretation of each case)

I would suggest using "`.volpkg` file", adding "file" to make it a noun. It should be more comprehensive.
I assume it is still a file (extension).

Reviewer: I like your suggestion, which I will modify to "`.volpkg` directory", since Volume Cartographer (and khartes) somewhat confusingly put file extensions on directory names.  Or if you think "folder" is more understandable than "directory", that is fine with me too.

Author: ```diff
- You might need to use `--volume` to specify your volume as well, if your `.volpkg` has more than one.
+ You might need to use `--volume` to specify your volume as well, if your `.volpkg` directory has more than one.
```

Directory is fine.

(though as always, remember to hit Ctrl-S periodically,
to save your changes to disk).

If you want to work on an area of the fragment that is
very steep, or even vertical, go to the Volumes Tab
to change the orientation of your current volume,
and then select your working region.

When you save your project, the new fragments
(and your changes)
will be saved along with the rest of the project;

> you will not need to re-import the fragment

Reviewer: ```suggestion
you will not need to re-import the fragment the
```

Reviewer: A rare case where I disagree with your suggestion.  I can't give a good grammatical reason, however; it's just a feeling.

Author: Okay.

Author: Unresolving comment. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

that when the tracking cursors are displayed, all the data slices are redrawn
every time the mouse is moved.

### Advanced topic: Live Updates

Khartes is designed to be interactive, so every time you make a change to a fragment, for instance, by
moving, adding, or deleting a node, the results of the change are shown instantly.  In the Data Slices,
the curve passing through the nodes is updated, and in the Fragment View, the surface texture is
updated.  Though these changes happen quickly, updates do require a certain amount of time, and when

> you have a large group of nodes you need to edit, you might find that updates will delay your work.

Reviewer: ```suggestion
you have a large group of nodes you need to edit, you might find that waiting for the updates
is delaying your work.
```

Author: Sure, looks fine.

Author: Unresolving comment. Modifications were confirmed to be okay (see https://github.com/KhartesViewer/khartes/pull/17#issuecomment-1793367671)

* This tutorial needs some drawings explaining the 3D positions of the Data Slices in the
scroll volume.

* This tutorial needs pictures of all the cursor shapes.

### License

Khartes is open-source code, licensed under the GPL 3.0 license.  See [LICENSE.txt](LICENSE.txt) for details.

### Contact

I prefer not to make my name and email address available on the public internet.
I can be reached on the [Vesuvius Challenge Discord server](https://discord.com/invite/6FgWYNjb4N)
under my user name, @khartes_chuck .

### Acknowledgements

I wrote khartes for my own gratification, and in the hope that it will prove useful;
I did not receive a grant or salary from any organization to develop this software.

However, I benefitted immensely from the [Vesuvius Challenge](https://www.scrollprize.org/), who
provided the following:
* The X-ray tomography data set, without which I could not have developed khartes (along with
making the data available under reasonable conditions,
they provided the technical infrastructure that made it possible to
download the data over the internet);
* Permission to use small excerpts from the data set in this README file and the associated videos;

> * A [lively community](https://discord.gg/V4fJhvtaQn) of fellow enthusiasts;

Reviewer: Thanks for updating the link!
